### ISSUE TITLE
Questions regarding Laravel and expose

### ISSUE THREAD

**Issue by jeff-h** [beyondcode/expose#63]
I have a couple of questions regarding using Laravel with expose:

* Laravel's `route()` and `url()` helpers both generate absolute URLs by default. This means all links in the site include the local URL, not the public "expose" URL. What's the best practice solution for resolving this situation when using expose?

* should the `.env` value for `APP_URL` still contain the local URL, or the public "expose" URL?

I've noticed a bunch of issues here already which relate to one of the above, but haven't really seen a definitive answer. I'm more than happy to create a PR adding some documentation around this, once I understand how things should be handled.

**Comment by marco-acorte** [comment]
it's an issue not strict related to expose (ngrok suffer the same issue).
the problem is all about relative/absolute urls,  according to laravel docs ( https://laravel.com/docs/7.x/helpers#method-route ) 

> By default, the route function generates an absolute URL. If you wish to generate a relative URL, you may pass false as the third argument:
> 
> `$url = route('routeName', ['id' => 1], false);`

**Comment by marco-acorte** [comment]
this route links (login/register) works normally and "exposed"
![immagine](https://user-images.githubusercontent.com/6375935/85272880-0ec62600-b47d-11ea-8a4e-8de78ed2e22b.png)

**Comment by jeff-h** [comment]
Thanks @marco-acorte for your help. Unfortunately my app is pretty large so it's not really an option to force relative URLs using the third argument of `route()`.

I have had some success with this in my `AppServiceProvider`:

```php
public function boot(Request $request)
{
    if ($request->server->has('HTTP_X_ORIGINAL_HOST')) {
        $this->app['url']->forceRootUrl($request->server->get('HTTP_X_FORWARDED_PROTO').'://'.$request->server->get('HTTP_X_ORIGINAL_HOST'));
    }
}
```

...but that feels like it could be a security issue (any proxy server could set those  headers) and it also doesn't feel like code that should be pushed to production.

**Comment by detygon** [comment]
@jeff-h Thanks for the code I just tried Expose and found this issue. For security concerns, maybe adding a check with `app()->environment(['local', 'testing'])` may help.

**Comment by mpociot** [comment]
I am unable to properly reproduce this issue.

I created a new Laravel project and created a new test route to my `web.php` routes file:

```php
Route::get('/test', function(){})->name('test');
```

In my `welcome.blade.php` file I then added links to these routes:

```
<div class="links">
    {{ url('test') }}<br/>
    {{ route('test') }}
</div>
```

Then I share the site running `expose share expose-example.test` and see the correct URLs:

![image](https://user-images.githubusercontent.com/804684/85294047-76d93400-b49e-11ea-8ad1-b9ca07511e79.png)

**Comment by jeff-h** [comment]
@mpociot aha! Thanks for that. It confirms my suspicion that there are only specific circumstances that cause the problem. Once I can allocate some time I will attempt to replicate your success with a fresh install, then see if I can isolate the problem from there.

Can I just confirm just one thing, which is whether you have made any changes to the config of your test install? Especially that you did not add any `TrustProxies` configuration?

**Comment by jeff-h** [comment]
Oh also, thinking more about it, probably the most significant question is: what server are you using for your test? My local is a virtual host nginx running behind an nginx reverse proxy. In other words, my Laravel app is available at something like https://localhost:8443, and I reverse-proxy that to https://myapp.test (on port 443).

**Comment by jeff-h** [comment]
Also @detygon, mind me asking what local server environment you're using?

**Comment by detygon** [comment]
@mpociot  Hi have you changed your `APP_URL` in the `.env` file ? Maybe that's what causing the issue. @jeff-h I use laradock for local dev.

**Comment by detygon** [comment]
It seems more like it is related to the dev environment since I also use nginx.

**Comment by mpociot** [comment]
> Oh also, thinking more about it, probably the most significant question is: what server are you using for your test? My local is a virtual host nginx running behind an nginx reverse proxy. In other words, my Laravel app is available at something like https://localhost:8443, and I reverse-proxy that to https://myapp.test (on port 443).

I'm using Laravel Valet, so there's not another reverse proxy in between. I also haven't changed any configuration of my Laravel app to make this work, no.

**Comment by clmntgr** [comment]
Having the same issue here, when I'm going to the sharedwithexpose.com url I'm redirected to my local URL.
I've updated my .env file with expose URL, and tried @jeff-h fix but didn't change anything.
When checking the dashboard I can see that all my request get a 301 redirect to the local url by nginx. I'm using Valet too.
Maybe I missed something

(Ngrok was working)

**Comment by mpociot** [comment]
The 301 redirect is most of the time coming from sharing a https secured url, but sharing it via http 

So you could try:

expose share https://your-local.url 

Instead

**Comment by clmntgr** [comment]
Sorry for that.

I can't reproduce this one so

**Comment by danydev** [comment]
Pretty sure all those kind of issues will go away when we'll have https://github.com/beyondcode/expose/issues/57
It's likely that the redirection your app is doing is based on the (wrong) Host header. As a counter-test you could try ngrok with and without the host rewrite feature

**Comment by jeff-h** [comment]
I think we're getting somewhere 🎉

In the [working example above](https://github.com/beyondcode/expose/issues/63#issuecomment-647525511), Laravel, when running under Valet, knows that the host is `http://expose-example.sharedwithexpose.com`.

If you check `http://127.0.0.1:4040` you can view the headers (fantastic feature btw 👍).
* the `host` header contains the **local** host name
* the only reference to the **remote** host name is in the `x-original-host` header

To my knowledge, Laravel knows nothing of this `x-original-host` header. My evidence is that it's totally ignored by Laravel.

**However**, the Valet Laravel driver does [some](https://github.com/laravel/valet/blob/bec26e2e3b4f977b035d46f6b2214393c365d573/server.php#L146) [magic](https://github.com/laravel/valet/blob/master/cli/drivers/LaravelValetDriver.php#L57) in order to convert the `x-original-host` header to a `x-forwarded-host` header and then finally to the `host` header (not disimilar to [my snippet above](https://github.com/beyondcode/expose/issues/63#issuecomment-647419530)).

Basically:
* your localhost needs to be given the host it expects (eg `local-site.test`)
* Laravel needs to be given the external host, so it can build URLs correctly (eg `external-site.sharedwithexpose.com`)
* in order for laravel to get this, either PHP in Laravel does the translation, or the local nginx/Apache needs to do it

**tldr;** In order for Laravel to work correctly within Expose, we must use a snippet of PHP code in Laravel, or some equivalent header-swapping cleverness in the local server config.

@mpociot please let me know your thoughts. If you think the above sounds right, I would like to shift the discussion a little onto what the best practice would be for this code snippet (and where best to put it), as well as what it would look like doing the header juggling in an nginx config. Hopefully then we can get some solutions for everyone with similar issues (hopefully even @danydev's issue).

**Comment by mpociot** [comment]
@jeff-h Maybe it would be enough if the expose server would add the `x-forwarded-host` header information as well, just like we add the x-original-host?

So it would be best if we don't have to touch Laravel at all (or any other application shared via expose). 
Symfony HTTP Foundation for example checks for the X-Forwarded-Host header to determine the correct request host (which again is used internally to build Laravel URLs): https://github.com/symfony/http-foundation/blob/71294564cb844f11d423f0d4cc292e52c7855269/Request.php#L1162

**Comment by jeff-h** [comment]
I agree it would be awesome to avoid a requirement for any changes in Laravel. I was wondering/hoping a solution might be found in Laravel's TrustProxies, but haven't used them before. Your Symfony reference above certainly sounds promising — can you think of some way I could test your `x-forwarded-host` header idea? If not, I am sure I could do it through my nginx config.

On a sidenote, unfortunately I'm pretty confident that web apps that don't leverage Symfony are not going to be straightforward (eg Drupal, at least v7 and earlier). However, Ngrok has at least paved the way for all this to a degree. I'm not sure if you'd want to go this way for your project but I'd be happy to attempt some brief documentation for getting specific frameworks up and running (at least Laravel and Drupal, as a start).

**Comment by jeff-h** [comment]
@mpociot Hooray! It worked. I added the following to my local nginx reverse proxy config:

```nginx
proxy_set_header  X-Forwarded-Host $http_x_original_host;
```

which takes the incoming header `x-original-host` from Expose, and adds a new header `x-forwarded-host` with the same value.

The final piece of the puzzle is to tell Laravel to trust this forwarded host header. In my test case I tell it to trust all proxies, although I will see if I can tighten that up:

```php
class TrustProxies extends Middleware
{
    /**
     * The trusted proxies for this application.
     *
     * @var array|string|null
     */
    protected $proxies = "*";
``` 

So if you were happy to have Expose add the `x-forwarded-host` header with the remote host in it, that would be great. It's correct, expected behaviour for a reverse proxy to add this header (and in fact, also the other `x-forwarded-*` headers), and when you think about it, arguably Expose *is* a fancy reverse proxy.

**Comment by mpociot** [comment]
I just added this header to expose

**Comment by harish81** [comment]
In case anyone need quick solution.
```php
public function boot()
{
    if(\App::environment()=='local' && isset($_SERVER['HTTP_X_ORIGINAL_HOST'])){
        $this->app['url']->forceRootUrl($_SERVER['REQUEST_SCHEME']."://".$_SERVER['HTTP_X_ORIGINAL_HOST']);
    }
}
```

> This will work for both `expose` and `ngrock`

**Comment by pxlrbt** [comment]
@mpociot
As a lot of people already had this issue, I think this should be mentioned somewhere in the docs. Maybe in the "Share your first site" section or maybe on a separate page with examples for different frameworks/CMS.

**Comment by dakira** [comment]
This is still a problem. Reproduce like this:

1. `laravel new mysite --jet --stack=livewire`
2. enable profile pictures in *config/jetstream.php*
3. Register an account and add a profile picture.
4. Share with expose and look at the profile picture url. It will be *http://mysite.test/...*

This is because the local storage disk adapters getUrl function will always take its url from the config setting `'filesystems.disks.public.url'` which uses `env('APP_URL').'/storage'`.

The following snippet in the boot() method of AppServiceProvider helps.

```php
if (app()->environment() === 'local' && isset($_SERVER['HTTP_X_ORIGINAL_HOST'])) {
    config(
        [
            'filesystems.disks.public.url' =>
                $_SERVER['HTTP_X_FORWARDED_PROTO']
                . '://' . $_SERVER['HTTP_X_ORIGINAL_HOST'] . '/storage',
        ]
    );
}
```

I don't see a way to fix this permanently in either Expose or Laravel. @mpociot

**Comment by ragingdave** [comment]
@mpociot Did this potentially get reverted around having the HOST header added? I am just attempting to expose a local artisan serve, and all my route helper calls are reflecting 127.0.0.1:8000 rather than the correct expose host

**Comment by glspdotnet** [comment]
Heya, @ragingdave. You might want to try the fix I posted [here](https://github.com/beyondcode/expose/issues/24#issuecomment-805396556) in another issue. Basically, you create a config file for Laravel's TrustedProxy dependency so that it knows to allow proxied requests from 127.0.0.1.

**Comment by amar-delve** [comment]
> Thanks @marco-acorte for your help. Unfortunately my app is pretty large so it's not really an option to force relative URLs using the third argument of `route()`.
> 
> I have had some success with this in my `AppServiceProvider`:
> 
> ```
> public function boot(Request $request)
> {
>     if ($request->server->has('HTTP_X_ORIGINAL_HOST')) {
>         $this->app['url']->forceRootUrl($request->server->get('HTTP_X_FORWARDED_PROTO').'://'.$request->server->get('HTTP_X_ORIGINAL_HOST'));
>     }
> }
> ```
> 
> ...but that feels like it could be a security issue (any proxy server could set those headers) and it also doesn't feel like code that should be pushed to production.

it will work for fixed host header attack issues?

**Comment by mityukov** [comment]
I'm experiencing the issue with built-in PHP server.

1. Install Laravel;
2. runt `php artisan serve`
3. `expose share http://localhost:8000`
